server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -146,7 +358,281 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
+        return actions;
+    }
+
+    /**
+     * Convenience method for filtering node allocation results by decider outcomes.
+     * @param deciderName The decider that is being checked
+     * @param outcome The outcome expected
+     * @return A predicate that returns true if the decision exists and matches the expected outcome, false otherwise.
+     */
+    private static Predicate<NodeAllocationResult> hasDeciderResult(String deciderName, Decision.Type outcome) {
+        return (nodeResult) -> nodeResult.getCanAllocateDecision()
+            .getDecisions()
+            .stream()
+            .anyMatch(decision -> deciderName.equals(decision.label()) && outcome == decision.type());
+    }
+
+    /**
+     * Generates a user action if a shard cannot be allocated anywhere because allocation is disabled for that shard
+     * @param indexMetadata from the index shard being diagnosed
+     * @param nodeAllocationResults allocation decision results for all nodes in the cluster.
+     * @result A list of user actions to take.
+     */
+    List<UserAction.Definition> checkIsAllocationDisabled(IndexMetadata indexMetadata, List<NodeAllocationResult> nodeAllocationResults) {
+        List<UserAction.Definition> actions = new ArrayList<>();
+        if (nodeAllocationResults.stream().allMatch(hasDeciderResult(EnableAllocationDecider.NAME, Decision.Type.NO))) {
+            // Check the routing settings for index
+            Settings indexSettings = indexMetadata.getSettings();
+            EnableAllocationDecider.Allocation indexLevelAllocation = INDEX_ROUTING_ALLOCATION_ENABLE_SETTING.get(indexSettings);
+            ClusterSettings clusterSettings = clusterService.getClusterSettings();
+            EnableAllocationDecider.Allocation clusterLevelAllocation = clusterSettings.get(
+                EnableAllocationDecider.CLUSTER_ROUTING_ALLOCATION_ENABLE_SETTING
+            );
+            if (EnableAllocationDecider.Allocation.ALL != indexLevelAllocation) {
+                // Index setting is not ALL
+                actions.add(ACTION_ENABLE_INDEX_ROUTING_ALLOCATION);
+            }
+            if (EnableAllocationDecider.Allocation.ALL != clusterLevelAllocation) {
+                // Cluster setting is not ALL
+                actions.add(ACTION_ENABLE_CLUSTER_ROUTING_ALLOCATION);
+            }
+        }
+        return actions;
+    }
+
+    /**
+     * Generates user actions for common problems that keep a shard from allocating to nodes in a data tier.
+     * @param indexMetadata Index metadata for the shard being diagnosed.
+     * @param nodeAllocationResults allocation decision results for all nodes in the cluster.
+     * @param clusterState the current cluster state.
+     * @return a list of user actions to take.
+     */
+    List<UserAction.Definition> checkDataTierRelatedIssues(
+        IndexMetadata indexMetadata,
+        List<NodeAllocationResult> nodeAllocationResults,
+        ClusterState clusterState
+    ) {
```

Comment:
Is there a way we can break this method up? It's very difficult to read as it's well over 100 lines long. Maybe we can put the `ShardsLimitAllocationDecider` branch into a method (as well as the other self-contained `if` statements under the top-level `else` clause?

Reply:
fixed in [f5bf091](https://github.com/elastic/elasticsearch/pull/85071/commits/f5bf0915a73f8f893111249d7e5985ae062fd1c5)